### PR TITLE
Fix input/output mismatch between TCS and TES

### PR DESCRIPTION
**Pull request by xuechen417**
There are two kind of mismatch between TCS and TES:
1. Lds offset mismatch: TCS writes outputs in a loop with loop index as the location offset while TES reads from the constant location. For example, TCS will write all 3 float values to LDS in order after loop unroll. After re-mapping, TES actually reads from location 0 but it expects the value of location 1. 
   
To address this issue, we need add the locations with dynamic location offset info to the TES input map. We modify the interface of `markGenericInputOutputUsage` to pass the dynamic location offset flag. Then, we initialize location map according to the flag. If the location map value is not invalid value, we will add them to TES input map in `updateInputLocInfoMapWithUnpack`. Besides, we refine the code of searching location maps in `PatchInOutImportExport::visitCallInst`. We should search the corresponding map based the call type, i.e., inputLocInfoMap, perPatchInputLocMap or perPrimitiveInputLocMap.

2. Location mapping mismatch: TCS packs outputs with the same location but TES maps inputs with same location with different components to continuous locations. 

To fix it and reduce LDS consumption, we will pack TES as what TCS does.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0c5daba45efc386f74a6d05e624cb3b5ed7596c7
<details><summary>CTS tests (Failed: 0/209336)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34737/69790 (49.8%)</li>
<li>Failed: 0/69790 (0.0%)</li>
<li>Not Supported: 35053/69790 (50.2%)</li><li>Warnings: 0/69790 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38622/69771 (55.4%)</li>
<li>Failed: 0/69771 (0.0%)</li>
<li>Not Supported: 31149/69771 (44.6%)</li><li>Warnings: 0/69771 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 38654/69775 (55.4%)</li>
<li>Failed: 0/69775 (0.0%)</li>
<li>Not Supported: 31121/69775 (44.6%)</li><li>Warnings: 0/69775 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by xuechen417** [comment]
V1: set the component for mapped TES input

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4a88e5c9c7350b1a3f5a514f17b7531742ba515b
<details><summary>CTS tests (Failed: 0/209353)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34737/69790 (49.8%)</li>
<li>Failed: 0/69790 (0.0%)</li>
<li>Not Supported: 35053/69790 (50.2%)</li><li>Warnings: 0/69790 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38620/69788 (55.3%)</li>
<li>Failed: 0/69788 (0.0%)</li>
<li>Not Supported: 31168/69788 (44.7%)</li><li>Warnings: 0/69788 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 38654/69775 (55.4%)</li>
<li>Failed: 0/69775 (0.0%)</li>
<li>Not Supported: 31121/69775 (44.6%)</li><li>Warnings: 0/69775 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/PatchResourceCollect.cpp [hunk 1]
```diff
@@ -2697,18 +2697,58 @@ void PatchResourceCollect::updateInputLocInfoMapWithUnpack() {
     }
   }
 
+  // Special processing for TES inputLocInfoMap and prePatchInputMap
```
Comment (nhaehnle) [inline lgc/patch/PatchResourceCollect.cpp:2700]: Typo: perPatchInputLocMap

File: lgc/builder/InOutBuilder.cpp [hunk 2]
```diff
@@ -417,23 +423,26 @@ void BuilderImpl::markGenericInputOutputUsage(bool isOutput, unsigned location,
...
+    unsigned initialVal = (isDynLocOffset && locationCount > 1) ? location : InvalidValue;
     // Non-GS-output case.
     if (inOutLocInfoMap) {
       for (unsigned i = startLocation; i < location + locationCount; ++i) {
         InOutLocationInfo origLocationInfo;
         origLocationInfo.setLocation(i);
         origLocationInfo.setComponent(inOutInfo.getComponent());
         auto &newLocationInfo = (*inOutLocInfoMap)[origLocationInfo];
-        newLocationInfo.setData(InvalidValue);
+        newLocationInfo.setData(initialVal);
```
Comment (nhaehnle) [inline lgc/builder/InOutBuilder.cpp:436]: Why isn't this offset somehow? That is to say, if locationCount > 1, why do all of the original locations get mapped to the same location here? I would have rather expected something like:
```
  newLocationInfo.setData(forceDynamicUse ? i : InvalidValue);
```

**Review by nhaehnle — COMMENTED** [review]
Thank you for the test case. The changes in PatchInOutImportExport seem like nice cleanups. PatchResourceCollect is always very difficult to judge, but I think it's okay.

**Comment by amdvlk-admin** [comment]
### Test summary for commit aadd08756ff6a88e7114e3c3bc20f440f5c8f952
<details><summary>CTS tests (Failed: 0/209351)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34735/69788 (49.8%)</li>
<li>Failed: 0/69788 (0.0%)</li>
<li>Not Supported: 35053/69788 (50.2%)</li><li>Warnings: 0/69788 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38620/69788 (55.3%)</li>
<li>Failed: 0/69788 (0.0%)</li>
<li>Not Supported: 31168/69788 (44.7%)</li><li>Warnings: 0/69788 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 38654/69775 (55.4%)</li>
<li>Failed: 0/69775 (0.0%)</li>
<li>Not Supported: 31121/69775 (44.6%)</li><li>Warnings: 0/69775 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/PatchResourceCollect.cpp (hunk 1 above)
Comment (xuechen417) [reply lgc/patch/PatchResourceCollect.cpp:2744]: Fixed.

File: lgc/builder/InOutBuilder.cpp (hunk 2 above)
Comment (xuechen417) [reply lgc/builder/InOutBuilder.cpp:436]: There is no special reason to map to the same location for dynamic location offset. I just want to make it different from InvalidValue in order to identify this location is caculated based on an offset later. Changed to your perfered shape.

**Comment by xuechen417** [comment]
V1: address the two comments.

**Comment by amdvlk-admin** [comment]
### Test summary for commit e46edd6ad03c98a937b31608212738dff3b23a4e
<details><summary>CTS tests (Failed: 0/209324)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34735/69779 (49.8%)</li>
<li>Failed: 0/69779 (0.0%)</li>
<li>Not Supported: 35044/69779 (50.2%)</li><li>Warnings: 0/69779 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38617/69779 (55.3%)</li>
<li>Failed: 0/69779 (0.0%)</li>
<li>Not Supported: 31162/69779 (44.7%)</li><li>Warnings: 0/69779 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 38651/69766 (55.4%)</li>
<li>Failed: 0/69766 (0.0%)</li>
<li>Not Supported: 31115/69766 (44.6%)</li><li>Warnings: 0/69766 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0742f03f2689b0325020ff9902852cca7fa77938
<details><summary>CTS tests (Failed: 0/209324)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34735/69779 (49.8%)</li>
<li>Failed: 0/69779 (0.0%)</li>
<li>Not Supported: 35044/69779 (50.2%)</li><li>Warnings: 0/69779 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38617/69779 (55.3%)</li>
<li>Failed: 0/69779 (0.0%)</li>
<li>Not Supported: 31162/69779 (44.7%)</li><li>Warnings: 0/69779 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 38651/69766 (55.4%)</li>
<li>Failed: 0/69766 (0.0%)</li>
<li>Not Supported: 31115/69766 (44.6%)</li><li>Warnings: 0/69766 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — DISMISSED** [review]
Thanks

**Comment by xuechen417** [comment]
V2: add a minor fix in PatchInOutImportExport. For checking inputLocInfoMap in unpacked patch, we should consider <location, component> since the same location inputs are compacted.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 79550912c19744ae53ce1636e9c4baa04115f4b3
<details><summary>CTS tests (Failed: 310/209040)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34734/69779 (49.8%)</li>
<li><details><summary>Failed: 1/69779 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.tessellation.shader_input_output.gl_position_vs_to_tcs
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 35044/69779 (50.2%)</li><li>Warnings: 0/69779 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38616/69495 (55.6%)</li>
<li><details><summary>Failed: 1/69495 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.tessellation.shader_input_output.gl_position_vs_to_tcs
Stack trace: Fail
```
</p></details></li>

<li>Not Supported: 30878/69495 (44.4%)</li><li>Warnings: 0/69495 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 38352/69766 (55.0%)</li>
<li><details><summary>Failed: 308/69766 (0.4%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.buffer.suballocation.storage_texel.storage.create.binding_aliased
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer.suballocation.transfer_src.uniform_texel.storage.index.create.binding_residency_aliased
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer_marker.graphics.external_host_mem.bottom_of_pipe.memory_dep.dispatch
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer_view.access.dedicated_alloc.buffer_view_memory_test_partial_offset0_with_buffer_dedicated_alloc_image_suballocated
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.copy_commands2.blit_image.all_formats.color.3d.a8b8g8r8_srgb_pack32.a8b8g8r8_srgb_pack32.general_general_linear_stripes_z
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.blit_image.all_formats.color.2d.r32_sfloat.e5b9g9r9_ufloat_pack32.optimal_general_nearest
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.blit_image.all_formats.color.2d.r4g4_unorm_pack8.r32_sfloat.general_general_nearest
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.blit_image.all_formats.generate_mipmaps.from_previous_level.layercount_6.r16g16b16a16_sfloat.optimal_optimal_nearest
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.depth_stencil_msaa_copy.whole.d16_unorm_s8_uint_general_general_S_2_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.depth_stencil_msaa_copy.whole.d32_sfloat_optimal_optimal_D_4_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.image_to_image.dimensions.src4096x6_dst4096x6.r32_sfloat.r8g8b8a8_unorm.optimal_general
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.resolve_image.whole_array_image_one_region.2_bit
Stack trace: Flake

FAILURE: dEQP-VK.api.fill_and_update_buffer.dedicated_alloc.fill_buffer_vk_whole_size_3_extra_bytes_offset_12
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.core.clear_color_image.2d.linear.single_layer.r16_unorm_clamp_input
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_attachment.multiple_layers.r16g16_uint_1x33
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.2d.linear.remaining_array_layers_twostep.r16g16b16a16_sint_1x33
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.2d.optimal.single_layer.b5g6r5_unorm_pack16_1x33_clamp_input_multiple_subresourcerange
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_depth_stencil_attachment.multiple_layers.d32_sfloat_s8_uint_33x128
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_depth_stencil_attachment.single_layer.d16_unorm_s8_uint_1x33
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.partial_clear_color_attachment.single_layer.r8g8_srgb_64x11_clamp_input
Stack trace: Flake
...
```
</p></details></li>

<li>Not Supported: 31106/69766 (44.6%)</li><li>Warnings: 0/69766 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1cf8a5d40a9066f9735a5f9d187225a840aca193
<details><summary>CTS tests (Failed: 0/209040)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 8.9, Gfx10</summary><ul>

<li>Passed: 34736/69779 (49.8%)</li>
<li>Failed: 0/69779 (0.0%)</li>
<li>Not Supported: 35043/69779 (50.2%)</li><li>Warnings: 0/69779 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38617/69495 (55.6%)</li>
<li>Failed: 0/69495 (0.0%)</li>
<li>Not Supported: 30878/69495 (44.4%)</li><li>Warnings: 0/69495 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 38651/69766 (55.4%)</li>
<li>Failed: 0/69766 (0.0%)</li>
<li>Not Supported: 31115/69766 (44.6%)</li><li>Warnings: 0/69766 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — DISMISSED** [review]

**Comment by xuechen417** [comment]
V3: Missing a map search statement in the last push.

**Comment by amdvlk-admin** [comment]
### Test summary for commit f398c30b13de17e2cb535f8e5dc2be0eecbc6189
<details><summary>CTS tests (Failed: 0/209324)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34735/69779 (49.8%)</li>
<li>Failed: 0/69779 (0.0%)</li>
<li>Not Supported: 35044/69779 (50.2%)</li><li>Warnings: 0/69779 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38617/69779 (55.3%)</li>
<li>Failed: 0/69779 (0.0%)</li>
<li>Not Supported: 31162/69779 (44.7%)</li><li>Warnings: 0/69779 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 38651/69766 (55.4%)</li>
<li>Failed: 0/69766 (0.0%)</li>
<li>Not Supported: 31115/69766 (44.6%)</li><li>Warnings: 0/69766 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit cac4be6a3b59039d654821ece0426c3f1bc65b07
<details><summary>CTS tests (Failed: 0/209324)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34735/69779 (49.8%)</li>
<li>Failed: 0/69779 (0.0%)</li>
<li>Not Supported: 35044/69779 (50.2%)</li><li>Warnings: 0/69779 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38617/69779 (55.3%)</li>
<li>Failed: 0/69779 (0.0%)</li>
<li>Not Supported: 31162/69779 (44.7%)</li><li>Warnings: 0/69779 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 38651/69766 (55.4%)</li>
<li>Failed: 0/69766 (0.0%)</li>
<li>Not Supported: 31115/69766 (44.6%)</li><li>Warnings: 0/69766 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit a49e7db2f17c4b03347ad1f44fe73e03a7f84bdc
<details><summary>CTS tests (Failed: 0/209040)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34735/69779 (49.8%)</li>
<li>Failed: 0/69779 (0.0%)</li>
<li>Not Supported: 35044/69779 (50.2%)</li><li>Warnings: 0/69779 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38617/69495 (55.6%)</li>
<li>Failed: 0/69495 (0.0%)</li>
<li>Not Supported: 30878/69495 (44.4%)</li><li>Warnings: 0/69495 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 38651/69766 (55.4%)</li>
<li>Failed: 0/69766 (0.0%)</li>
<li>Not Supported: 31115/69766 (44.6%)</li><li>Warnings: 0/69766 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]

**Pull request merged**